Added STMDA and STMDB instructions semantics

NeatMonster opened the pull request:
STM and STMIB are already implemented. STMDA and STMDB simply call the helper
function _STM(). This function is modified to support decrementing the address.

See also ARM Architecture Reference Manual.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/trailofbits/manticore/1245)
<!-- Reviewable:end -->

disconnect3d commented on manticore/core/cpu/arm.py:
```diff
@@ -1066,14 +1066,21 @@ def _STM(cpu, insn_id, base, regs):
-            address += 4
+            if insn_id in (cs.arm.ARM_INS_STM, cs.arm.ARM_INS_STMIB):
+                address += 4
+            if insn_id in (cs.arm.ARM_INS_STMDA, cs.arm.ARM_INS_STMDB):
```
```suggestion
            elif insn_id in (cs.arm.ARM_INS_STMDA, cs.arm.ARM_INS_STMDB):
```

NeatMonster replied:
I've refactored the `_LDM` and `_STM` methods. Hopefully they are easier to understand now.

disconnect3d commented on manticore/core/cpu/arm.py:
```diff
@@ -1066,14 +1066,21 @@ def _STM(cpu, insn_id, base, regs):
         address = base.read()
         if insn_id == cs.arm.ARM_INS_STMIB:
             address += 4
+        if insn_id == cs.arm.ARM_INS_STMDB:
```
```suggestion
        elif insn_id == cs.arm.ARM_INS_STMDB:
```

NeatMonster replied:
I've refactored the `_LDM` and `_STM` methods. Hopefully they are easier to understand now.

disconnect3d commented on manticore/core/cpu/arm.py:
```diff
@@ -1066,14 +1066,21 @@ def _STM(cpu, insn_id, base, regs):
         # Undo the last addition if we're in STMIB
         if insn_id == cs.arm.ARM_INS_STMIB:
             address -= 4
+        if insn_id == cs.arm.ARM_INS_STMDB:
```
```suggestion
        elif insn_id == cs.arm.ARM_INS_STMDB:
```

NeatMonster replied:
I've refactored the `_LDM` and `_STM` methods. Hopefully they are easier to understand now.

NeatMonster commented:
As mentioned in the commit message, I've refactored LDM/STM-related code, added docstrings and unit tests. More specifically:

* check if the S bit is set (usermode), which is not supported yet
* refactor the _LDM() and _STM() functions to be easier to understand
* if PC is specified in the registers list of LDM, change mode accordingly
* added the missing LDM* and STM* instructions semantics
* added docstring to _LDM() and _STM()
* added unit tests for all the instructions (+ check writeback)
